sktree/tree/_unsup_criterion.pyx

```diff
@@ -136,35 +218,223 @@ cdef class TwoMeans(UnsupervisedCriterion):
             The new ending position for which to move samples from the right
             child to the left child.
         """
-        # need to implement
-        pass
+        cdef SIZE_t pos = self.pos
+        cdef SIZE_t end = self.end
+
+        cdef const SIZE_t[:] sample_indices = self.sample_indices
+        cdef const DOUBLE_t[:] sample_weight = self.sample_weight
+
+        cdef SIZE_t i
+        cdef SIZE_t p
+        cdef SIZE_t k
+        cdef DOUBLE_t w = 1.0
+
+        # Update statistics up to new_pos
+        #
+        # Given that
+        #   sum_left[x] +  sum_right[x] = sum_total[x]
+        # and that sum_total is known, we are going to update
+        # sum_left from the direction that require the least amount
+        # of computations, i.e. from pos to new_pos or from end to new_po.
+        if (new_pos - pos) <= (end - new_pos):
+            for p in range(pos, new_pos):
+                i = sample_indices[p]
+
+                if sample_weight is not None:
+                    w = sample_weight[i]
+
+                # TODO: sum_left updated
+                for k in range(self.n_outputs):
```

Comment:
Remove notions of `n_outputs` as we discussed. Unless you have questions, lmk.

Reply:
Thank you!